### PR TITLE
API Update to behat 3

### PR DESCRIPTION
**Pull request by tractorcow**
Fixes https://github.com/silverstripe/silverstripe-behat-extension/issues/62

requires:

 - [x] https://github.com/silverstripe/silverstripe-testsession/pull/47 (high priority) and a tag 3.0.0-alpha1. Note: 2.0.0 of testsession has never completed alpha, maybe tag that an branch out to 2 at the same time?
 - [x] https://github.com/silverstripe/silverstripe-config/pull/9 (high priority, may need a tag)
 
 Then merge this PR. Subsequently, we should work on getting framework behat tests green. May need to tag a 3.0.0-alpha1 of behat-extension also. 
 
 - [ ] https://github.com/silverstripe/silverstripe-framework/pull/6830 - get framework behat tests passing to merge this.
 
 Once framework tests pass:
 
  - [ ] https://github.com/silverstripe/silverstripe-cms/pull/1801
  - [ ] https://github.com/silverstripe/silverstripe-installer/pull/172 (required for asset-admin)
  - [ ] https://github.com/silverstripe/silverstripe-asset-admin/pull/443
  
  Siteconfig module isn't yet upgraded, but is lower priority.
 
 
Rough overview of changes:

 - `@retry` tag can mark any scenario / feature as supporting retry. Any failed step will retry up to 3 times (once per second). This means no more `and I wait 1 second` garbage.
 - behat yml files are composable. Add `@<modulename>` and behat will automatically hunt down the module, load the root `behat.yml` in the module-root, and merge it with the site-root `behat.yml`.
 - Class name guessing just doesn't happen any more. It was always wrong and dumb. you need to specify the classes either in behat.yml, or via `--namespace` when doing `behat --init`.
 - Behat tests all run fine locally, but CI and phpunit tests need review. They are likely unstable until I get to tweak them.
 - A large number of behat injection vars have been MOVED to suite-specific configs. That means you define these on the individual context service definitions, NOT on the extension. This means that per-module config is more flexible.
 - SS_BASE_URL is now mandatory, and is used to set the mink base_url param internally.
 - Since contexts are now yml-assigned, the old "base" fixture structure from before doesn't work very well. In place, I've created a "SilverStripeContext" abstract class which must be implemented and named `FeatureContext` on EVERY test suite. This will hold the mink-specific information, so that other contexts can access this data. This is the `main context`. See `MainContextAwareTrait`

### CODE REVIEW

**Comment by tractorcow** [comment]
cc @chillu

File: composer.json [hunk 1]
```diff
@@ -1,53 +1,60 @@
...
+  ],
+  "require": {
+    "php": ">=5.6",
+    "phpunit/phpunit": "^5.7",
+    "behat/behat": "^3.2",
+    "behat/mink": "^1.7",
+    "behat/mink-extension": "^2.1",
+    "behat/mink-selenium2-driver": "^1.3",
+    "symfony/dom-crawler": "^3",
+    "silverstripe/testsession": "^2",
```
Comment (sminnee) [inline composer.json]: This should be `2.0.0-alpha6`

Comment (tractorcow) [inline composer.json]: Yes good idea.

**Comment by tractorcow** [comment]
Rebased and updated.

**Comment by sminnee** [comment]
This looks reasonable and I'm going to merge for the purposes of releasing an alpha1 of this package. I'll also create a `2` branch so that behat2 can be maintained while there are still modules using it.

**Pull request merged**